### ISSUE TITLE
IHL Internet Header Length unclear

### ISSUE THREAD

**Issue by Stargateur** [bestouff/pktparse-rs#14]
According to what I read, IHL minimum value is `5` and max is `15` cause 4 bits. If I understand the purpose of this `<< 2` shift is to multiply by 4 the number cause it's represent the number of 32-bit words. But this look very wrong to do that cause it's not documented and it's very unexpected in my opinion cause it's stored in a `u8` not a `usize`, this would also disallow to let this value store in one octet. In my opinion we should remove this shift or document it.

https://github.com/bestouff/pktparse-rs/blob/ebd2e18701b319fb76a21aaf9960bf84988d21c7/src/ipv4.rs#L59
https://github.com/bestouff/pktparse-rs/blob/ebd2e18701b319fb76a21aaf9960bf84988d21c7/src/ipv4.rs#L111

More [doc](https://www.freesoft.org/CIE/RFC/791/12.htm) and [`wiki`](https://en.wikipedia.org/wiki/IPv4#IHL)

---

Related: Should we check the value ?
Related: BTW, we don't handle ipv4 options

**Comment by bestouff** [comment]
Hi,
I don't get it. If the max size is 15 32-bits words, that means once shifted by 2 it's 60 which still fits an `u8`, no need for an `usize` for that.
But then I agree with you the shift isn't documented which is wrong. So, barring my possible misunderstanding of the issue, I see 2 ways out:
- we stop shifting, but probably requires a new major release because it'll break some downstream users
- we don't touch the code and just document the behavior

I'd favor the second solution but if we can discuss this with some of our users (@kpcyrd ?) and agree on something I'm all ears.

**Comment by Stargateur** [comment]
> I don't get it. If the max size is 15 32-bits words, that means once shifted by 2 it's 60 which still fits an u8, no need for an usize for that.

the point is `usize` exist to represent a `size` since the shift is here to transform it to a size in bytes/octets (and probably so use it for indexing an array) it's only logical to use the type for that in rust and that is `usize`. (hint user)

> we don't touch the code and just document the behavior

I think it's clearly the less painful solution, personally I don't use this field so I don't care.

**Comment by kpcyrd** [comment]
I'm not sure what's the best option, I'm only using this field for printing so all options work for me :) I've asked a friend and he suggested removing the shift.

**Comment by bestouff** [comment]
Do I make that a breaking change ? I'm a bit torn on this.

**Comment by kpcyrd** [comment]
I'd be fine with a breaking change, but I can see your concern. Your choice!

**Comment by Stargateur** [comment]
I just realize I was not clear with:

> this would also disallow to let this value store in one octet.

My point was if in future we want to transform the struct to be write in one block If we keep this shift it's impossible cause we could not stock it in 4 bits.

For example [bit-struct](https://github.com/andrewgazelka/bit-struct) crate I discover recently one could define the Ipv4Header struct

```rust
bit_struct! {
    struct IPv4Header(u8) {
        version: u4,
        ihl: u4,
        // etc...
    }
}
```

Thus I don't know if this is a good idea to do it, but be able to do it in the future would be nice. (I also have some idea about IpProtocol change)

**Comment by bestouff** [comment]
I released 0.7.1 with the change; if anyone reports this breaks something I'll yank it and republish as 0.8.0
